teps/0060-remote-resource-resolution.md

```diff
@@ -242,207 +251,533 @@ users can leverage those shared pipelines.
+resolution. In most cases this will be limited to `GET`, `LIST` and
+`UPDATE` permissions on `ResourceRequests` and their `status`
+subresource. The "in-cluster" resolver will need permissions to `GET`
+and `LIST` `Tasks`, `ClusterTasks` and `Pipelines` as well. Depending on
+requirements of each, some resolvers may need `GET` on `ConfigMaps` or
+`Secrets` in the `tekton-remote-resolution` namespace as well.
+
+A resolver is only ultimately responsible for updating the `status.data`
+and `status.annotations` field of a `ResourceRequest`. The
+`ResourceRequest` lifecycle controller is responsibile for marking
+resource requests completed or timing them out. A resolver _may_ mark a
+`ResourceRequest` as failed with an accompanying error and reason.
+
+#### In-Cluster Resolver
+
+The "in-cluster" resolver will support looking up `Tasks`,
+`ClusterTasks` and `Pipelines` from the cluster it resides in. It will
+mimic Tekton Pipelines' existing built-in support for these resources.
+The parameters for this `ResourceRequest` will look like:
+
+```yaml
+kind: Task
+name: foo
+```
+
+#### Bundle Resolver
+
+The "bundle" resolver will support looking up resources from Tekton
+Bundles. This will mirror the existing support in Tekton Pipelines. In
+addition it will verify a bundle before returning it if configured to do
+so. The precise approach to verifying a bundle will be based on the
+decisions made in
+[TEP-0091](https://github.com/tektoncd/community/pull/537).
+
+The supported paramaters for a "bundle" `ResourceRequest` will be:
+
+```yaml
+image_url: gcr.io/tekton-releases/catalog/upstream/golang-fuzz:0.1
+name: golang-fuzz
+signer: cosign # TBD, see TEP-0091
+```
+
+When the bundle is successfully resolved the `ResourceRequest` will be
+updated with both the resource contents and an annotation with the
+digest of the fetched image.
+
+#### Git Resolver
+
+The "git" resolver will support fetching resources from git
+repositories. This is new functionality that is not supported by Tekton
+Pipelines currently. In addition to basic support for fetching files
+from git this resolver will also need to support quite rich
+configuration (which can be developed over time):
+
+- Allow-lists for specific repo urls, branches and commits.
+- Settings for limiting access to certain repos by namespace.
+- Timeout settings for git operations.
+- Proxy settings (equivalent to `HTTP_PROXY`, `HTTPS_PROXY` environment
+  variables).
+- Path filtering such that only specific directories and file paths
+  within a repo can be sourced.
+- Verification of fetched files, following any approach decided in
+  [TEP-0091](https://github.com/tektoncd/community/pull/537).
+
+Since clones of large repos can be slow, and not all providers support
+features like sparse checkout, the "git" resolver will implement caching
+as a high priority.
+
+The parameters for a "git" `ResourceRequest` will initially look as
+follows:
```

Comment:
I think there are several distinct ideas here that are worth separating out for discussion purposes:

1. Support provider-specific optimizations when requesting resources (e.g. use github's api to read a file if possible rather than cloning an entire repo).
2. Create provider-specific resolvers.
3. Allow users to pick which provider is used in their `taskRef` or `pipelineRef`.

Let's look at each of these individually:

#### Support provider-specific optimizations

I agree with this one but am not sure it necessarily implies per-provider resolvers or exposing the provider as user-facing config (vs operator- or admin-facing).

#### Users should specify the provider in their `taskRef` or `pipelineRef`

My assumption here is that the git repo provider is likely to be extremely static information - I don't think a repository is going to hop from being a github repo to being a gitlab repo to being a gerrit repo within a short amount of time. Given the static nature of the information I think allowing users to specify the provider of a given git repo with every taskrun or pipelinerun would end up being redundant.

In addition to that I'm not totally convinced that this is configuration you want to put in the hands of end-users. Operators or administrators, definitely, but not necessarily end-users. That's not to discount the idea entirely - just that I think we should figure out some use cases for making this configurable by "everybody".

So instead what I'd suggest is that we make this operator-level configuration supplied to the git resolver:

```yaml
allowed-repositories:

- name: tekton-catalog
  repository_url: https://github.com/tektoncd/catalog.git
  provider: github

- name: app-team-1-private-repo
  repository_url: git@code.enterprise-foo.net:app-team-1
  provider: github-enterprise
```

By default the git resolver can treat every git repo the same. When a repository is attributed to a specific provider though there can be code paths in the git resolver that can use provider-specific optimizations to fetch a resource.

But!  We might not have code paths for every git repo provider under the sun which leads on to...

#### Provider-Specific Resolvers

I think one of the nice things about the syntax / design as proposed is that we can provide a general purpose `git` resolver out of the box with optimizations for the most popular providers. Someone else can in turn implement resolvers for their own `niche-git-provider` which use specific optimizations.

Reply:
> Support provider-specific optimizations when requesting resources (e.g. use github's api to read a file if possible rather than cloning an entire repo).

Note that using APIs instead of git directly (GitHub API) has its set of limitation too (rate limits, …). Also are we making the assumption that you need to do a git clone with all path and all history to get a specific commit and path from a git repository ? Because I don't think this assumption is completely valid, there is way to "clone" only the thing you are interested into (without grabing the full history of the repo nor the full content of it at a given point)

Reply:
Right, I think this is also a fair stance to take! A lot of this is going to have to shake out at implementation time I think.

In terms of the TEP I think we can look at this from the perspectives of different user profiles:
- Users who submit TaskRuns and PipelineRuns shouldn't have to worry about any of the nuanced detail of interacting with the repo.
- Operators _maybe_ care about some of the specifics so they can configure those things once.
- Developers of `git` Resolvers have to really care about it and figure it all out in code, test cases and abstractions.